File: app/src/main/java/sourcecontrol/core/repository/Repository.java
package sourcecontrol.core.repository;

import java.nio.file.Path;
import java.util.Optional;

import sourcecontrol.core.objects.GitObject;
import sourcecontrol.core.objects.ObjectStore;
import sourcecontrol.exceptions.RepositoryException;

public interface Repository {
    /**
     * Initialize a new repository at the given path
     */
    void init(Path path) throws RepositoryException;

    /**
     * Get the working directory path
     */
    Path getWorkingDirectory();

    /**
     * Get the .git directory path
     */
    Path getGitDirectory();

    /**
     * Get the object store
     */
    ObjectStore getObjectStore();

    /**
     * Read an object from the repository
     */
    Optional<GitObject> readObject(String sha) throws RepositoryException;

    /**
     * Write an object to the repository
     */
    String writeObject(GitObject object) throws RepositoryException;

    /**
     * Check if repository exists at path
     */
    static boolean exists(Path path) {
        return path.resolve(".git").toFile().exists();
    }
}

File: sourcecontrol/pkg/repository/sourcerepo/repository.go
package sourcerepo

import (
	"github.com/utkarsh5026/SourceControl/pkg/objects"
	"github.com/utkarsh5026/SourceControl/pkg/repository/scpath"
	"github.com/utkarsh5026/SourceControl/pkg/store"
)

// Repository defines the interface for Git repository operations.
// It provides access to the repository's working directory, git directory,
// and object storage.
type Repository interface {
	// Initialize creates a new repository at the given path
	Initialize(path scpath.RepositoryPath) error

	// WorkingDirectory returns the path to the repository's working directory
	WorkingDirectory() scpath.RepositoryPath

	// SourceDirectory returns the path to the .source directory (equivalent to .git)
	SourceDirectory() scpath.WorkingPath

	// ObjectStore returns the object store for this repository
	ObjectStore() store.ObjectStore

	// ReadObject reads a Git object by its SHA-1 hash
	ReadObject(hash objects.ObjectHash) (objects.BaseObject, error)

	// WriteObject writes a Git object to the repository
	WriteObject(obj objects.BaseObject) (objects.ObjectHash, error)

	// Exists checks if a repository exists at the working directory
	Exists() (bool, error)
}
